### PR TITLE
Load split DWARF

### LINKED ISSUE #1805: Allow loading DWARF information from the separate files
Many Linux distributions now ship the DWARF debugging information separately from the binaries itself in a packages with `-debuginfo` or similar prefixes.

See:
- https://wiki.gentoo.org/wiki/Debugging
- https://wiki.debian.org/DebugPackage
- https://docs.fedoraproject.org/en-US/packaging-guidelines/Debuginfo/
- https://fedoraproject.org/wiki/Releases/36/ChangeSet#MinGW_debug_symbols_location_change
- https://fedoraproject.org/wiki/Changes/F36MingwDebugLocation

Curiously, the heap parsing code in `librz/core/linux_heap.c` already supports those splitdebug symbols for GLIBC symbols parsing.

### PR DESCRIPTION
**Pull request by b1llow**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [ ] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [ ] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

...

**Test plan**

...

**Closing issues**

Closes https://github.com/rizinorg/rizin/issues/1805

### CODE REVIEW

File: librz/core/cbin.c [hunk 1]
```diff
@@ -626,17 +626,30 @@ RZ_API bool rz_core_bin_apply_main(RzCore *r, RzBinFile *binfile, bool va) {
...
 
+RZ_API RzBinDWARF *rz_core_bin_dwarf(RzCore *core, RzBinFile *binfile) {
+	RzBinDWARFOption opt = {
+		.line_mask = RZ_BIN_DWARF_LINE_INFO_MASK_LINES,
+		.flags = RZ_BIN_DWARF_ALL - RZ_BIN_DWARF_LOC
+	};
+	RzBinDWARF *dw = rz_bin_dwarf_from_file(binfile, &opt);
+	if (!dw) {
+		const char *filepath = rz_config_get(core->config, "bin.dbginfo.filepath");
+		if (!filepath) {
```
Comment (XVilka) [inline librz/core/cbin.c]: Please support standard paths searching for Debian, Gentoo, etc

**Review by XVilka — CHANGES_REQUESTED** [review]
We need also a small test for splitdebug

**Comment by ret2libc** [comment]
Is this about loading DWARF information from a separate file??? That would be awesome <3

**Comment by XVilka** [comment]
@imbillow any updates on this one?

**Comment by b1llow** [comment]
not yet

**Comment by b1llow** [comment]
@XVilka

Do I need to do the debuginfod integration in this PR, or should I merge it first and leave the debuginfod part for the next?

**Comment by XVilka** [comment]
Better to do debuginfod support in a separate PR. Let's focus on simple splitdebug here.

File: librz/bin/dwarf/str.c [hunk 2]
```diff
@@ -28,14 +19,24 @@ RZ_IPI RzBinDwarfStr *RzBinDwarfStr_new(RZ_NONNULL RZ_OWN RzBinEndianReader *rea
...
 	rz_return_val_if_fail(bf, NULL);
-	RzBinEndianReader *r = RzBinEndianReader_from_file(bf, ".debug_str");
+	RzBinEndianReader *r = RzBinEndianReader_from_file(
+		bf, ".debug_str", is_dwo);
 	RET_NULL_IF_FAIL(r);
-	return RzBinDwarfStr_new(r);
+	return rz_bin_dwarf_str_new(r);
+}
+
+RZ_API void rz_bin_dwarf_str_free(RzBinDwarfStr *str) {
```
Comment (XVilka) [inline librz/bin/dwarf/str.c]: `RZ_NULLABLE`

File: librz/bin/dwarf/str_offsets.c [hunk 3]
```diff
@@ -27,39 +19,34 @@ RZ_IPI RzBinDwarfStrOffsets *RzBinDwarfStrOffsets_new(RzBinEndianReader *reader)
...
-	RzBinEndianReader *r = RzBinEndianReader_from_file(bf, ".debug_str_offsets");
+	RzBinEndianReader *r = RzBinEndianReader_from_file(
+		bf, ".debug_str_offsets", is_dwo);
 	RET_NULL_IF_FAIL(r);
-	return RzBinDwarfStrOffsets_new(r);
+	return rz_bin_dwarf_str_offsets_from_buf(r);
 }
-
-RZ_IPI char *RzBinDwarfStrOffsets_get(RzBinDwarfStr *str, RzBinDwarfStrOffsets *str_offsets, ut64 base, ut64 index) {
+RZ_API void rz_bin_dwarf_str_offsets_free(RzBinDwarfStrOffsets *str_offsets) {
```
Comment (XVilka) [inline librz/bin/dwarf/str_offsets.c]: `RZ_NULLABLE`

File: librz/core/cbin.c [hunk 4]
```diff
@@ -626,13 +626,25 @@ RZ_API bool rz_core_bin_apply_main(RzCore *r, RzBinFile *binfile, bool va) {
 	return true;
 }
 
+static inline RzBinDWARF *rz_core_bin_dwarf(RzCore *core, RzBinFile *binfile) {
```
Comment (XVilka) [inline librz/core/cbin.c]: Maybe because it's inline, no need for `rz_` prefix?

**Review by XVilka — CHANGES_REQUESTED** [review]

File: librz/core/cconfig.c [hunk 5]
```diff
@@ -3230,6 +3230,9 @@ RZ_API int rz_core_config_init(RzCore *core) {
 	SETI("bin.baddr", -1, "Base address of the binary");
 	SETI("bin.laddr", 0, "Base address for loading library ('*.so')");
 	SETCB("bin.dbginfo", "true", &cb_bindbginfo, "Load debug information at startup if available");
+	SETCB("bin.dbginfo.dwo_path", "", NULL, "Load separate debug information (DWARF) file if available");
+	SETCB("bin.dbginfo.debug_file_directory", "/usr/lib/debug", NULL,
```
Comment (XVilka) [inline librz/core/cconfig.c:3234]: Does it support multiple paths?

**Review by XVilka — CHANGES_REQUESTED** [review]
The test failing. Please also rebase on top of the latest `dev`

File: librz/core/cconfig.c (hunk 5 above)
Comment (b1llow) [reply librz/core/cconfig.c:3234]: Yes, multiple paths can be separated by `;`.

File: librz/bin/dwarf/str_offsets.c [hunk 6]
```diff
@@ -27,39 +19,34 @@ RZ_IPI RzBinDwarfStrOffsets *RzBinDwarfStrOffsets_new(RzBinEndianReader *reader)
...
+	if (!str_offsets) {
+		return;
+	}
+	rz_vector_free(str_offsets->offsets);
+	RzBinEndianReader_free(str_offsets->reader);
+	free(str_offsets);
+}
+RZ_API RZ_BORROW const char *rz_bin_dwarf_str_offsets_get(
```
Comment (XVilka) [inline librz/bin/dwarf/str_offsets.c:44]: Missing empty line

**Review by XVilka — CHANGES_REQUESTED** [review]
Test fails: https://github.com/rizinorg/rizin/actions/runs/6594200551/job/17917793126?pr=3833#step:12:268

**Review by XVilka — APPROVED** [review]

**Pull request merged**